jax/_src/prng.py

```diff
@@ -0,0 +1,471 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from functools import partial
+from typing import Callable, Iterator, NamedTuple
+import warnings
+
+import numpy as np
+
+from jax import lax
+from jax import core
+from jax import numpy as jnp
+from jax import tree_util
+from jax._src.api import jit, vmap
+from jax.lib import xla_bridge
+from jax.lib import xla_client
+from jax.lib import cuda_prng
+from jax.interpreters import batching
+from jax.interpreters import xla
+from jax._src.pprint_util import pp, vcat
+from jax._src.util import prod
+
+
+UINT_DTYPES = {
+    8: jnp.uint8, 16: jnp.uint16, 32: jnp.uint32, 64: jnp.uint64}  # type: ignore[has-type]
+
+# -- PRNG implementation interface --
+
+class PRNGImpl(NamedTuple):
+  key_shape: core.Shape
+  seed: Callable
+  split: Callable
+  random_bits: Callable
+  fold_in: Callable
+
+  def pprint(self):
+    return (pp(self.__class__.__name__) >> pp(':')) + vcat([
+        pp(k) >> pp(' = ') >> pp(v) for k, v in self._asdict().items()
+    ]).indent(2)
+
+
+# -- PRNG key arrays --
+
+def _is_prng_key_data(impl, keys: jnp.ndarray) -> bool:
+  ndim = len(impl.key_shape)
+  try:
+    return (keys.ndim >= 1 and
+            keys.shape[-ndim:] == impl.key_shape and
+            keys.dtype == np.uint32)
+  except AttributeError:
+    return False
+
+@tree_util.register_pytree_node_class
+class PRNGKeyArray:
+  """An array whose elements are PRNG keys.
+
+  This class lifts the definition of a PRNG, provided in the form of a
+  ``PRNGImpl``, into an array-like pytree class. Instances of this
+  class behave like an array whose base elements are keys, hiding the
+  fact that keys are typically arrays (of ``uint32`` dtype) themselves.
+
+  PRNGKeyArrays are also restricted relative to JAX arrays in that
+  they do not expose arithmetic operations. They instead expose
+  wrapper methods around the PRNG implementation functions (``split``,
+  ``random_bits``, ``fold_in``).
+  """
+
+  impl: PRNGImpl
```

Comment:
It's a natural question! I kicked around an idea like this earlier on.

I'm choosing to wrap-and-delegate because it directly expresses what I'm trying to enforce. `PRNGKeyArray` is a class made to adapt a PRNG implementation, which is expressed on an individual key, into an "array" of keys. It's an internal definition. We have no need in jax for a hierarchy of behaviors around it, let alone to offer that anyone outside try to do so.

If you're writing custom PRNGs from outside of jax, you're welcome to supply the PRNG interface however you like. It can be a member of whatever fun type hierarchy you've set up in your code, or just a plain record of four functions. You'll notice it need not even be a proper `prng.PRNGImpl`, but rather anything that duck-types as one. Hopefully that's plenty of flexibility, less for us to document, and less for anyone to learn.

To your specific points:

> * simplify your flatten and unflatten methods by removing `self.impl`,
> * simplify your implementation by removing `PRNGImpl`,

These seem like minimal savings, so I'm not concerned.

> * simplify type annotations,

How so?

> * make it easier to have a hierarchy of behavior if the need arises,

I don't foresee such a need. We can always revisit if/when it comes up.

> * and make it easier for users to define their own `PRNGKeyArray`.

I don't want them to!

What do you think? As someone who might use this, do these choices get in your way at all?

Reply:
Thanks for the exhaustive reply.

> `PRNGKeyArray` is a class made to adapt a PRNG implementation, which is expressed on an individual key, into an "array" of keys. It's an internal definition, not part of the API, and we have no need in jax for a hierarchy of behaviors around it.

`PRNGKeyArray` is a member of the union `KeyArray`, which is passed to the public functions in `random.py`.  So, `PRNGKeyArray` is part of the API, isn't it?

Did you mean `PRNGImpl` is not part of the API?  That appears to be private.  I guess my point is that it could just be folded into PRNGKeyArray since it doesn't appear that you use it on its own anywhere.  I do understand your point that you want to treat `PRNGKeyArray` as a wrapper.

> we have no need in jax for a hierarchy of behaviors around it.

Fair enough.

> simplify type annotations,
> How so?

The functions in `PRNGImpl` are annotated as `Callable`. It would be nice to just define them as abstract methods so that they have full signatures.  I guess I don't see the point of using function-valued members that you set instead of abstract methods.

> make it easier for users to define their own PRNGKeyArray.
> I don't want them to!

One possible user-defined PRNGKeyArray subclasss  you might find is a Haiku `HaikuKeyArray`.  It would simplify Haiku code to not have to extract keys from a `PRNGKeySequence`, and instead use the `HaikuKeyArray`, which would be passable to all the JAX random functions.

Why make life harder for users?

> do these choices get in your way at all?

They do not!  So I guess it doesn't matter.  I just thought it would be nicer to have one class rather than two, and to use abstract methods rather than function-valued members.

Reply:
> Thanks for the exhaustive reply.

Of course – thanks for the same. I better understand your points now. It seems like there are pros and cons either way. I went ahead with this PR as is, but we can always adapt in the future, and it will be useful to have this discussion to refer back to.

> `PRNGKeyArray` is a member of the union `KeyArray`, which is passed to the public functions in `random.py`. So, `PRNGKeyArray` is part of the API, isn't it? How should a user annotate the generator that's produced by `seed_with_impl` and that's passed around to user functions?

You're right – the type is publicly visible. What I had in mind was: users are not meant to instantiate this class directly, nor to interact with with it other than by (a) the array-like methods (which I want no one to override) and (b) passing it to `random` functions. It's supposed to simply look like an array for the most part. Good correction to my unqualified remark regardless.

> About `PRNGImpl`, my point is that it could just be folded into PRNGKeyArray since it doesn't appear that you use it on its own anywhere except to produce a `PRNGKeyArray`. I do understand your point that you want to treat `PRNGKeyArray` as a wrapper, but it only ever wraps one thing, and the thing it wraps is only ever used within the wrapper.

I see now. Yes, that would save some indirection. On the other hand, PRNG implementations would involve slightly more wrapping/unwrapping boilerplate that the class currently handles.

> They do not!

Excellent. That's the most important part to me. Everything else can be revised later if/as needed.